orderer/common/localconfig/config.go

```diff
@@ -284,6 +297,50 @@ var Defaults = TopLevel{
 // a struct suitable for config use, returning error on failure.
 func Load() (*TopLevel, error) {
 	config := viper.New()
+
+	coreconfig.InitViper(config, "orderer")
+	config.SetEnvPrefix(Prefix)
+	config.AutomaticEnv()
+	replacer := strings.NewReplacer(".", "_")
+	config.SetEnvKeyReplacer(replacer)
+
+	if err := config.ReadInConfig(); err != nil {
+		return nil, fmt.Errorf("Error reading configuration: %s", err)
+	}
+
+	uconf := &TopLevel{}
+	cfgCache.mux.Lock()
+	serializedConf, ok := cfgCache.cache[config.ConfigFileUsed()]
+	if !ok {
+		err := viperutil.EnhancedExactUnmarshal(config, uconf)
+		if err != nil {
+			return nil, fmt.Errorf("Error unmarshaling config into struct: %s", err)
+		}
+		uconf.completeInitialization(filepath.Dir(config.ConfigFileUsed()))
+
+		serializedConf, err = json.Marshal(uconf)
+		if err != nil {
+			return nil, err
+		}
+		cfgCache.cache[config.ConfigFileUsed()] = serializedConf
+		cfgCache.mux.Unlock()
+
+		return uconf, nil
+	}
+	cfgCache.mux.Unlock()
+
+	err := json.Unmarshal(serializedConf, uconf)
+	if err != nil {
+		return nil, err
+	}
+	return uconf, nil
+}
+
+// loadWithoutCache performs the same functionality as Load, but without using configCache.
+// This makes sure configs that being override by environment variable are read correctly.
+func loadWithoutCache() (*TopLevel, error) {
```

Comment:
If this is hung off of the cache, then the tests simply get to create a new instance of the cache to test the functionality.

Reply:
I'm not quite clear about this one. **loadWithoutCache** function loads config from viperutil.EnhancedExactUnmarshal, but not from configCache. Should it not be hanging off the cache?

Reply:
I'm suggesting you reorder things a bit. Once you have a cache "object" that memoizes the value you've read from the config, the exported function can use that directly to get (and cache) the value.

Within the tests (where you don't want caching), you simply instantiate a new instance of the cache and use it to load the config. The memoization will be done relative to whatever the current environment variables are.